### ISSUE TITLE
`ViewableData::ThemeDir()` broken

### ISSUE THREAD

**Issue by dhensby** [silverstripe/silverstripe-framework#6030]
## Acceptance Criteria

 * `$ThemeDir` is restored to templates
 * The value of ThemeDir depends on the template it is being used within
 * The value will the root-relative URL (i.e. starts with "/") of the theme folder, as processed by ResourceURLGenerator.
 * A template string such as `<img src="$ThemeDir/images/something.png">` will reference images/something.png within the theme.

Exclusions:

 * Direct access to ThemeDir from ViewableData isn't necessary (so it can be implemented as a template global instead if preferred)

----

`$ThemeDir` is used in templates to pull in assets. This is currently used by https://github.com/silverstripe-themes/silverstripe-simple to pull in a JS file.

However this is now broken after moving to cascading themes.

# PRs

 - [ ] https://github.com/silverstripe/silverstripe-framework/pull/7626

**Comment by sminnee** [comment]
In SS4, `$ThemeDir` should probably resolve to the base path of the module or theme that contains the template in question.

**Comment by marcinqunabu** [comment]
$ThemeDir in SS4 gives '/mysite' somehow, not the theme folder itself.

**Comment by sminnee** [comment]
Yeah in SS4 there's less distinction between modules and themes, so /mysite is a theme, as is every other module, as well as the actual theme(s) selected.

Because of this, it doesn't make sense to say "where's the theme?" Instead it's better to say "what is the root of *this* theme?"

In other words, the value of ThemeDir would depend upon the template in question being rendered.

ThemeDir should probably also be deprecated in favour of the package resource management system proposed in https://github.com/silverstripe/silverstripe-framework/issues/6243, once it exists

**Comment by sminnee** [comment]
It seem that since this ticket has been raised, ThemeDir has been removed, which isn't really a fix.

My view is that ThemeDir's value should depend upon which template it's being used within.

 * /themes/something/templates/Page.ss - ThemeDir returns /themes/something
 * /themes/other/templates/Layout/Page.ss - ThemeDir returns /themes/other
 * /vendor/silverstripe/simple/templates/Layout/Page.ss - ThemeDir returns /resources/silverstripe/simple/

I'm not sure if this is a bugfix (4.0.x) or a new feature (4.x) as it would help ease upgrade pain. It's been raised a number of times on silverstripe-users

**Comment by tractorcow** [comment]
Maintaining the script filename as a part of the template parser could be alot of added complexity; The template parser converts all scripts to PHP before execution.

The new replacement is the `$resourcePath` / `$resourceURL` helpers which are more useful, in that they allow you to not only refer to specific resources within the same theme, but perform resource-to-url transformation, which $ThemePath ignores.

I would argue that any place that $ThemePath is used it wouldn't be any less correct to hard-code the name of the theme the template is saved in. E.g. if you are creating a template for `silverstripe-themes/simple`, then it's assuredly going to be in the path `themes/simple`.

If you are creating a new module theme (which can be placed in vendor) then $ThemePath won't be useful at all, since it would include `vendor` and is thus un-routable anyway, and you should use `$resourceUrl` instead.

To address this issue I would suggest to ensure adequate upgrading notes have been written, and that there is a direct upgrade path for all prior usages of $ThemeDir that previously existed.

**Comment by sminnee** [comment]
> I would argue that any place that $ThemePath is used it wouldn't be any less correct to hard-code the name of the theme the template is saved in. E.g. if you are creating a template for silverstripe-themes/simple, then it's assuredly going to be in the path themes/simple.

This isn't true, because we're trying to decouple URL from file location. The path **needs** to be passed through the URLResourceGenerator. Notably, it's intended to return a URL not a file path.

**Comment by tractorcow** [comment]
>  Notably, it's intended to return a URL not a file path.

I don't think it's that obvious, but sure I could accept that. In any case, what's proposed is a lot of extra complexity that only saves users from switching to an existing and superior upgrade substitution.

**Comment by sminnee** [comment]
I generally agree that ThemeResource would be a better API, so our efforts should probably go into getting people onto that more easily. There will be a lot of `$ThemeDir`s in template.

Options:

 * docs improvement
 * shim method that throws a warning/error
 * upgrader extension to rewrite templates

**Comment by tractorcow** [comment]
Oh look I was wrong.

```php
$template = $res['template'];
```

So I guess that "extra complexity" is already there somehow.

In that case, just add some new SSTemplateParser code to substitute themedir from that path?

**Comment by sminnee** [comment]
Ooh, right, it doesn't look like we currently have a `ThemeResource` template helper. So we should probably add that? Here's some community-contributed code:

```
public function ThemedResource($thing){
      $base = Director::baseURL();
      $path = ThemeResourceLoader::inst()->findThemedResource($thing,  SSViewer::get_themes());
      $ret = '';
      if($path === NULL){
          // Could not find rewource
          $ret = "[not found] $thing";
      }else{
          $ret = $base.$path;
      }
      return "$ret";
  }
```

**Comment by sminnee** [comment]
Other note: it may be better to provide ThemeDir and/or ThemeResource as a template global helper rather than a ViewableData method (since ViewableData is already too god-objecty), and the API that we're looking to preserve is access to ThemeDir from within templates (that's the use-case that users are tripping over)

**Comment by tractorcow** [comment]
The code I was referring to was in ModuleResourceLoader, since it's per-module not per-configured-themes.

```
public static function get_template_global_variables()
    {
        return [
            'resourcePath',
            'resourceURL'
        ];
    }
```

**Comment by tractorcow** [comment]
I would suggest to limit any additional API to that which maintained usages of the prior 3.x template vars.

**Comment by sminnee** [comment]
Oh this is relevant too: https://silverstripe-users.slackarchive.io/-/search-themedir/page-1

**Comment by sminnee** [comment]
Right so new syntax is:

```
<img src=$resourceURL(my/theme: images/foo.png)>
```

**Comment by tractorcow** [comment]
Yes that's correct. :)

We can fix $ThemeDir probably in 4.0.1 depending on priority.

Should we mirror $resourcePath / $resourceURL and add $themedResourcePath / $themedResourceURL ?

**Comment by sminnee** [comment]
If I were designing the API now, I probably wouldn’t include resourcePath as it’s not really the kind of data a template render can make use of.

So, I’m a little reluctant about themedResourcePath, but if you see a value in it I don’t have strong views. Putting URL on the end for naming consistency makes sense though.

I’ll break that piece out into a separate ticket, too.

**Comment by chillu** [comment]
`$ThemeDir` seems like a frequently used API, so let's fix this in 4.0.1 (as opposed to 4.1.0) - assigning milestone.

**Comment by sminnee** [comment]
I've drafted some ACs above — can @chillu and @tractorcow :+1: them?

**Comment by sminnee** [comment]
OK I've broken out the follow-on work as #7592. It's not critical for 4.1; I've marked it as priority medium.

**Comment by tractorcow** [comment]
> So, I’m a little reluctant about themedResourcePath, but if you see a value in it I don’t have strong views. Putting URL on the end for naming consistency makes sense though.

It'll be useful outside of templates, when trying to get file contents to be manipulated, for example. :)

**Comment by tractorcow** [comment]
Fine with me @chillu

**Comment by dhensby** [comment]
> I'm not sure if this is a bugfix (4.0.x) or a new feature (4.x) as it would help ease upgrade pain. It's been raised a number of times on silverstripe-users

I think the most important thing is that there's some clear upgrade guidance for this. I have been playing around with upgrading a site of my own from 3.6 -> 4.0 and there isn't clear guidance on how to pull in a themed resource (but there is for how to pull in a resource from a module).

In the end I replaced use of `$this->ThemeDir()` with `ThemeResourceLoader::inst()->getPath('default')` in PHP. This feels dirty as the path on the server may not always map to the url.

For templates I replaced `$ThemeDir` for importing CSS with `<% require themedCSS("css/main.css") %>`, which seemed much more reasonable.
